Use a more descriptive error message on an exception check validation failure

ast-hugger opened the pull request:
#### fabbe3fe6abdb696c8a430322a2d32565a17e72d
<pre>
Use a more descriptive error message on an exception check validation failure
<a href="https://bugs.webkit.org/show_bug.cgi?id=297604">https://bugs.webkit.org/show_bug.cgi?id=297604</a>
<a href="https://rdar.apple.com/158698040">rdar://158698040</a>

Reviewed by Mark Lam.

An exception check validation failure in `VM::verifyExceptionCheckNeedIsSatisfied` is
currently triggered by `RELEASE_ASSERT(!m_needExceptionCheck)`. The assert condition is
invariably false because it appears in an `if` block guarded by `m_needExceptionCheck`.
Written this way it is not obvious that the failure is not a violation of internal VM
state consistency. It might even suggest that there used to be logic in this code,
now missing, that was expected to reset the flag to false.

The patch changes the assert to clarify that the unconditional failure is intentional
and explicitly state the reason.

Canonical link: <a href="https://commits.webkit.org/298929@main">https://commits.webkit.org/298929@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/813522ea195be0e98df49eb813f3a59a257d6747

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/117128 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36776 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/27379 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/123194 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/69126 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/4b21ff69-2486-4d14-9e10-52e451993826) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/37491 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/45383 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88905 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/43616 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/7603012a-da9e-4154-b20d-5b49c1d3e9a1) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120061 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29859 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105045 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69397 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28919 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/23155 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/66848 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/109255 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99250 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/23321 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/126335 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/115637 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44018 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33068 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97578 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/44372 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101269 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97377 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24803 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42720 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20651 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/40398 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43893 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/49624 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/144337 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/43360 "Built successfully") | | [💥 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/37146 "An unexpected error occured. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/46706 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45070 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/runtime/VM.cpp:
```diff
@@ -1426,7 +1426,7 @@ void VM::verifyExceptionCheckNeedIsSatisfied(unsigned recursionDepth, ExceptionE
 
         dataLog(out.toCString());
-        RELEASE_ASSERT(!m_needExceptionCheck);
+        RELEASE_ASSERT_WITH_MESSAGE(false, "exception check validation failed");
```
Why not `RELEASE_ASSERT_NOT_REACHED_WITH_MESSAGE()`?

ast-hugger replied:
I grepped for "RELEASE_ASSERT_WITH_MESSAGE(false" and saw several precedents. I'll change to this one, thanks!

MenloDorian reviewed (approved):
r=me

ast-hugger committed:
Use a more descriptive error message on an exception check validation failure
https://bugs.webkit.org/show_bug.cgi?id=297604
rdar://158698040

Reviewed by Mark Lam.

An exception check validation failure in `VM::verifyExceptionCheckNeedIsSatisfied` is
currently triggered by `RELEASE_ASSERT(!m_needExceptionCheck)`. The assert condition is
invariably false because it appears in an `if` block guarded by `m_needExceptionCheck`.
Written this way it is not obvious that the failure is not a violation of internal VM
state consistency. It might even suggest that there used to be logic in this code,
now missing, that was expected to reset the flag to false.

The patch changes the assert to clarify that the unconditional failure is intentional
and explicitly state the reason.

Canonical link: https://commits.webkit.org/298929@main
